src/string_id.h

```diff
@@ -69,69 +107,96 @@ template<typename T>
 class string_id
 {
     public:
+        template<class U>
+        friend class string_id;
         using This = string_id<T>;
 
-        /**
-         * Forwarding constructor, forwards any parameter to the std::string
-         * constructor to create the id string. This allows plain C-strings,
-         * and std::strings to be used.
-         */
-        // Beautiful C++11: enable_if makes sure that S is always something that can be used to constructor
-        // a std::string, otherwise a "no matching function to call..." error is generated.
+        template<unsigned N>
+        explicit constexpr string_id( const char ( &str )[N] ) : _hash( Hash::CompileTime::fnv1_64( str ) ),
+            _char_ptr( str ), _is_char_ptr( true ) {}
+
+        template<typename U>
+        string_id( const string_id<U> &other ) : _hash( other.hash() ), _char_ptr( other._char_ptr ),
+            _is_char_ptr( other._is_char_ptr ) {}
+
         template<typename S, class = typename
-                 std::enable_if< std::is_convertible<S, std::string >::value>::type >
-        explicit string_id( S &&
-                            id ) : _id( std::forward<S>( id ) ), _cid( INVALID_CID ), _version( INVALID_VERSION ) {}
+                 std::enable_if< std::is_convertible< S, std::string >::value >::type
+                 >
+        explicit string_id( S && id ) {
+            auto str = std::make_unique<std::string>( std::forward<S>( id ) );
+            uint64_t hash = Hash::RunTime::fnv1_64( str->c_str(), str->size() );
+            _hash = hash;
+
+            auto insertion = cached_strings().try_emplace( hash, std::move( str ) );
```

Comment:
What happens if we do get a hash collision?

Reply:
As it currently stands - str() will point at a wrong string, but as stated, this can be solved.

Reply:
I think you can change the `std::map` to a `std::set` which takes a hash-string pair as the key, so when the hashes are the same it automatically falls back to string comparison, which should avoid the hash collision.